### ISSUE TITLE
[DOC] Userguide adaptation to Decoder object

### ISSUE THREAD

**Issue by thomasbazeille** [nilearn/nilearn#2415]
The Metaestimator Decoder is not well described in the user guide. For example (but it's more general than this) the part "introduction to decoding" still has example code with sklearn functions, which seems out of sync with the example it draws figures from (plot_decoding_tutorial.py)

With a release including the Decoder object, it seems very important that the most didactic parts of Nilearn include clear examples using it.

Side point to discuss : I think we should keep some examples / userguide sections describing how to use sklearn directly for decoding; But then we need to find how to articulate clearly both together.

Any thoughts ?

**Comment by jeromedockes** [comment]
ThomasBaz <notifications@github.com> writes:


> Side point to discuss : I think we should keep some examples / userguide sections describing how to use sklearn directly for decoding; But then we need to find how to articulate clearly both together.

maybe by placing those that do not use the decoder in the "07_advanced"
section of the examples?

**Comment by bthirion** [comment]
Yes. It should be almost hidden: this is not the first thing our users should try.

**Comment by emdupre** [comment]
Is this a blocker for the 0.7.x release series ?

**Comment by thomasbazeille** [comment]
Not sure, I planned to focus on it during the sprint so maybe an alpha
release in the mean time is no problem since old documentation is still
valid. But at the same time if we get some exposure with your communication
we could want to advertise Decoder a lot !

Le mar. 28 avr. 2020 à 16:46, Elizabeth DuPre <notifications@github.com> a
écrit :

**Comment by thomasbazeille** [comment]
On the Userguide :

2.1  We should include an explicit reference to the new example on how to fit first-level GLM before decoding in 2.1.1.1 section of userguide. 
And to be exemplary we should maybe change subsequent examples using Haxby raw data to use beta-maps instead but this is costly.Thoughts ?

2.1.2 / 2.1.3 are the primary area that needs a lot of reworking. I think it's valid to present concepts of what is a prediction model, cross validation, a scorer, and chance, and then feature selection but the code should definitely be using the Decoder object, and I'm not sure we should mention sklearn objects that early.

I think section 2.1.4 and introduction of sklearn objects should not be done here but in a later session


Section 2.2 should be adapted well, I think it's the right time to distinguish Regression, Classification and different linear models, but it should be done with the Decoder object.

We should also mention here the specificity of the Decoder object in terms of bagging models learned on different folds together instead of refitting with best parameters. It would be good as well to illustrate how to do a confusion matrix with the Decoder object.

2.3 / 2.4 / 2.5 I think this is roughly the right time to speak about fREM, Space Net and Searchlight. I didn't check if Searchlight was up-to-date with Decoder object.

Maybe we should add a 2.6 section repackaging some content of actual session 2.1.2 and 2.1.4 on how to do decoding with scikit-learn estimators directly so that there is a guide to more autonomy and flexibility for people who need it.

Examples :

Example 9.1.4. "A introduction tutorial to fMRI decoding" uses Decoder and should be synced with new section 2.1.2/2.1.3

9.3.8 on feature selection  is *not adapted yet* and should use Decoder and be synced with 2.1.3 or 2.2
9.3.9 on ROI masking is *not adapted yet* and should use Decoder and be synced with 2.2
9.3.10 on using various estimators is adapted and  2.2 should be synced with it.
9.3.11 on Regression is adapted already and 2.2 should be synced with it

9.8.7 on fitting a GLM + Decoding should be included at the beginning of 9.3 section or even in 9.1 Tutorial examples 

fREM examples (9.3.2, 9.3.3) Space-Net example (9.3.4) Searchlight (9.3.6) are up-to-date but should be put after more basic example if sphinx allow

Some others I'm less sure : 
- 9.3.5 about multi-class strategies could be adapted to be used in decoder
- 9.3.7 is made useless by the Decoder I guess
- 9.3.12 I don't get what it is trying to show from a pedagogical point of view ? 

Also we should maybe have a full sklearn example in advanced section 9.8 for more edgy usecases linked to 2.6

Sorry this is a lot

**Comment by bthirion** [comment]
Is GLM application on Haxby really costly ?

> bagging models learned on different folds together instead of refitting with best parameters
I think this is quite advanced  and technical. Should be presented not too early.

**Comment by bthirion** [comment]
> Maybe we should add a 2.6 section repackaging some content of actual session 2.1.2 and 2.1.4 on how to do decoding with scikit-learn estimators directly so that there is a guide to more autonomy and flexibility for people who need it.

Yes, it makes sense to me.

**Comment by thomasbazeille** [comment]
> Is GLM application on Haxby really costly ?

I mean it's not from *very* costly computationally but for now the fetcher is on raw images and we don't want to fetch, then apply a GLM, in *every* example. It would add confusion and useless overhead. 
So either : 
- we do it once, and add a fetcher for these beta-maps which is a bit costly in dev and maintenance but principled
- either we male a clear disclaimer : "You should use beta-maps as in example ... but for simplicity we use raw data"

> > bagging models learned on different folds together instead of refitting with best parameters
> 
> I think this is quite advanced and technical. Should be presented not too early.

Agreed but it should be clear somewhere obvious since we want users to understand what they are doing no ?

**Comment by bthirion** [comment]
For the examples, I agree with you. But I don't think that the ordering is really important.

**Comment by bthirion** [comment]
> 9.3.7 is made useless by the Decoder I guess

I think that it is always important to emphasize that. This is really educational.

**Comment by bthirion** [comment]
> 9.3.12 I don't get what it is trying to show from a pedagogical point of view ?

That inference ond ecoder maps is hard and ill-posed. I think that this is an important message. But I agree that the text should be rewritten.

**Comment by bthirion** [comment]
> I mean it's not _very_ costly computationally but for now the fetcher is on raw images and we don't want to fetch, then apply a GLM, in _every_ example. It would add confusion and useless overhead.
> So either :
> 
>     * we do it once, and add a fetcher for these beta-maps which is a bit costly in dev and maintenance but principled
> 
>     * we make a clear disclaimer : "You should use beta-maps as in example ... but for simplicity we use raw data"

OK, I think I misunderstood you.
It's fine to use Hacby without GLM fitting, as long as cross validation is across sessions. Still I would ass "you may prefer to fit a GLM first and perform decoding on the resulting beta maps, as in example XXX ".

**Comment by SylvainTakerkart** [comment]
Hi, I took a quick look...

One suggestion: change the title of the 9.3.2 example from "FREM on Jimura dataset" to something like "Regression with FREM" (in order to match and complement the title of 9.3.3 which is "Decoding with FREM"

**Comment by bthirion** [comment]
> One suggestion: change the title of the 9.3.2 example from "FREM on Jimura dataset" to something like "Regression with FREM" (in order to match and complement the title of 9.3.3 which is "Decoding with FREM"

+1

**Comment by SylvainTakerkart** [comment]
Hi! Copying my feedback from discord to here... Several items...

---

So, I ran the haxby Decoder example (plot_haxby_different_estimators.py)... The decoder.cvscores seems to be organized by class (shoe, scissors etc.), which I think might be confusing for the users, no? (coz we always have to explain to beginners that it's not like the GLM, you don't get "one map, or one result, per class/experimental condition"; and this exhibits the opposite...). Thoughts?

---

I have the same feeling with plot_haxby_full_analysis.py ... what most people do, is that they compare binary or multiclass decoding performances across regions; but this example also splits it by class, which might be misleading to beginner users (as I mentioned above)... maybe you could exhibit this behavior in one example (plot_haxby_different_estimators.py) and exhibit another behavior (without splitting by class) in plot_haxby_full_analysis.py to better match what (I believe) most people do in the literature when comparing decoding across ROIs?

---

Several other points, somehow summarized from our discussion...
- how about using / displaying a confusion matrix instead?
- with regards to having one coef image attribute per class (in the Decoder object) in the multiclass case; in fact, in the binary classification case, you do not have one coef image attribute, correct? (I have not checked the latter)

**Comment by bthirion** [comment]
Regarding results per class: If I'm not mistaken, this is equivalent to looking at the diagonal of the confusion matrix, so I think it's not crasy, but indeed, we should make this explicit. 

> maybe you could exhibit this behavior in one example (plot_haxby_different_estimators.py) and exhibit another behavior (without splitting by class) in plot_haxby_full_analysis.py to better match what (I believe) most people do in the literature when comparing decoding across ROIs?

I think  I agree with that.
One thing: I don't want to encourage people to compare accuracies across regions, because such accuracy differences are not tested for significance...
As a side remark: this is actually part of the SPM dogma: you cannot conclude about regions differences by doing local (ROI-based or mass-univariate) analyses.

**Comment by bthirion** [comment]
>     * with regards to having one coef image attribute per class (in the Decoder object) in the multiclass case; in fact, in the binary classification case, you do not have one coef image attribute, correct? (I have not checked the latter)

Actually in that case the two maps would be opposite to each other; we produce only one.

**Comment by thomasbazeille** [comment]
> what most people to is that they compare binary or multiclass decoding performances across regions; but this example also splits it by class
Agreed !  Maybe we can showcase how to do both in  plot_haxby_full_analysis.py ?


> in the binary case, you don't have "one coef image per class"! correct?
I think you actually do, maybe opposite to one another here.

> how about using / displaying a confusion matrix instead?
I think we can totally display it in several examples and Doc section when relevant but i'm a bit against adding it as an attribute for Decoder by default. Anyone else has thoughts ? (@jeromedockes )

**Comment by SylvainTakerkart** [comment]
> One thing: I don't want to encourage people to compare accuracies across regions, because such accuracy differences are not tested for significance...

I agree! but this example (plot_haxby_full_analysis.py) definetely does encourage people to compare decoding across regions...

**Comment by SylvainTakerkart** [comment]
> >``` in fact, in the binary classification case, you do not have one coef image attribute, correct? (I have not checked the latter)
> > ```
> 
> Actually in that case the two maps would be opposite to each other; we produce only one.

OK, but, let's say, in the case of "shoe vs house" classification , labeling one map as "shoe" and the other as "house" is, I think misleading... This is probably not a good solution, but maybe they should be labeled "shoe vs house" and "house vs shoe", similarly to "shoe - house" and "house - shoe" for univariate contrasts...

It is the single-class labeling that bothers me...

**Comment by bthirion** [comment]
In the multi-class case, the label should be something like 'house vs others'.

**Comment by thomasbazeille** [comment]
example 9.8.7 (now transfered to decoding section) should be modified with Decoder as well

**Comment by thomasbazeille** [comment]
Due to the removal of scikit-learn objects, the introduction of Chance Level notion in section 2.1.2.2.4. and how to determine it using dummy estimator that was there before is not very compatible API anymore.

Maybe we should accept Decoder(estimator="dummy") to measure chance quickly with friendly API  ?

**Comment by bthirion** [comment]
+1 for `Decoder(estimator="dummy")`

**Comment by thomasbazeille** [comment]
@tbng the open PR on documentation is linked just above. Can we split the examples to be adapted

I'll take care of the ones most related with docs : 
- Example 9.1.4. "A introduction tutorial to fMRI decoding" uses Decoder and should be synced with new section 2.1.2/2.1.3
- 9.8.7 on fitting a GLM + Decoding not adapted yet and should use Decode, should be included at the beginning of 9.3 section or even in 9.1 Tutorial examples
- And try to make a full sklearn example in advanced section 9.8 for more edgy usecases linked to 2.6

Can you take care of :
- 9.3.8 on feature selection is not adapted yet and should use Decoder and be synced with 2.1.3 or 2.2
- 9.3.9 on ROI masking is not adapted yet and should use Decoder and be synced with 2.2
- 9.3.10 on using various estimators and  9.3.11 on Regression should be done already, can you check whether they're fine ?

Then we have to decide what to do with 9.3.5 about multi-class strategies, 9.3.7 and 9.3.12

**Comment by tbng** [comment]
On 16.09.2020 03:29, ThomasBaz wrote:
>@tbng the open PR on documentation is linked just above. Can we split the examples to be adapted

>Can you take care of :
>- 9.3.8 on feature selection is not adapted yet and should use Decoder and be synced with 2.1.3 or 2.2
>- 9.3.9 on ROI masking is not adapted yet and should use Decoder and be synced with 2.2
>- 9.3.10 on using various estimators and  9.3.11 on Regression should be done already, can you check whether they're fine ?

Thanks for making it clear. So I guess I would open a new PR for improvements of these examples?

**Comment by bthirion** [comment]
Indeed, can you do it quickly  ?
Thx

**Comment by thomasbazeille** [comment]
> To elaborate working on a branch of your PR ([gh-2476](https://github.com/nilearn/nilearn/pull/2476)) is recommended or should I create a branch from nilearn/master?

Separate PR for the example sounds perfect as its independent from the doc

**Comment by tbng** [comment]
>     * 9.3.8 on feature selection is not adapted yet and should use Decoder and be synced with 2.1.3 or 2.2
> 
>     * 9.3.9 on ROI masking is not adapted yet and should use Decoder and be synced with 2.2
> 
>     * 9.3.10 on using various estimators and  9.3.11 on Regression should be done already, can you check whether they're fine ?
> 
> 
> Then we have to decide what to do with 9.3.5 about multi-class strategies, 9.3.7 and 9.3.12

Sorry I don't find section 9 on Nilearn documentation website. Can you provide the link to Python file for those example instead?

**Comment by thomasbazeille** [comment]
Closing as most of this was adress, the remainer is listed in #2564
